### ISSUE TITLE
Multi-GPU performance improvement for ASR in ESPnet1 by DDP

### ISSUE THREAD

**Issue by lazykyama** [espnet/espnet#4377]
**Describe the bug**

Dear ESPnet team,

This issue is a proposal to improve the multi-GPU training performance for **v1** ASR.
I'm filing this because I didn't find any other issues and/or PRs except for https://github.com/espnet/espnet/issues/3583 as far as I checked. (Sorry in advance if I overlooked any other similar issues/PRs and/or should discuss on https://github.com/espnet/espnet/issues/3583.)
I have already made a basic DDP implementation for a verification as mentioned later, and can send a PR.
Here, I would like to discuss with you if this feature is required at first.
After that, also would like to talk about if the current modifications are acceptable, etc.

As discussed in https://github.com/espnet/espnet/issues/3583, it looks like ASR training script in v1 has a large bottleneck in multi-GPU case.
As far as I investigated, it's caused by thread-based multi-GPU implementation.
According to a profiling result by Nsight Systems, GPU computation occasionally stopped even though actual training codes don't contain any blockers, and GPU has enough capability to continue calculation.
Also, during some GPUs stopped, it looks like only one GPU could run simultaneously, and the GPU that could run was determined in a round-robin fashion.
Therefore, I guess that this is caused by GIL.

To verify if my guess is correct, I made several changes for ASR v1 training codes to support PyTorch's DDP.
As a result, 8GPU case can achieve x2 faster compared to non-DDP version (=current implementation).
Performance data is below.
Note that 8GPU case includes several times validations because 2500 iters mean around 10 epochs in my setting, and only `train-clean-100` dataset is used for this measurement to avoid too long training time.

| num of GPUs | num of iters | current impl [sec.] | DDP [sec.] | 
|:--:|:--:|:--:|:--:|
| 1 | 2000 | 1270.37 | 1299.95 | 
| 8 | 2500 | 5327.21 | 2470.36 | 

And, throughput roughly calculated from the result above is below. (bs=32 is assumed.)

| num of GPUs | current impl [samples / sec.] | DDP [samples / sec.] | DDP vs current impl |
|:--:|:--:|:--:|:--:|
| 1 | 201.5160937 | 196.9306512 | x0.98 |
| 8 | 480.5517335 | 1036.286209 | x2.16 |

**Basic environments:**

I did an early investigation and PoC implementation on DGX A100 (80GB memory version) and within a NGC's PyTorch docker container.
Detailed information is below.

 - OS information: `Linux 4c930d24d0ff 5.4.0-110-generic #124-Ubuntu SMP Thu Apr 14 19:46:19 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`
 - python version: Python 3.8.5
 - espnet version: espnet 0.10.4
 - Git hash: eee2566f2395d676e3978376315a4cd116a4e335
   - Commit date: Wed Nov 10 17:23:24 2021 +0900
   - We need to test transducer recipe. So, recipe is only v.202204 (commit hash: 48c351057df1ff9d5605d9cb4361cf1f216cf533).
 - pytorch version pytorch 1.8.0a0+1606899
   - Base container image: https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/rel_21-12.html#rel_21-12

**Environments from `torch.utils.collect_env`:**

(If necessary, please let me know. I will add these information later.)

**Task information:**

 - Task: ASR
 - Recipe: ibrispeech
 - ESPnet1

**To Reproduce**

1. `cd egs/librispeech/asr1`
2. run `run.sh` until stage 3 to setup everything,
3. execute `asr_train.py` like below.

```
python /workspace/espnet/espnet/bin/asr_train.py \
    --epochs 10 \
    --minibatches -1 \
    --config conf/tuning/transducer/train_conformer-rnn_transducer.yaml \
    --preprocess-conf conf/specaug.yaml \
    --ngpu 8 \
    --backend pytorch \
    --outdir exp/train_100_pytorch_train_conformer-rnn_transducer_specaug/results \
    --tensorboard-dir tensorboard/train_100_pytorch_train_conformer-rnn_transducer_specaug \
    --debugmode 1 \
    --dict /data/dataset/asr/librispeech/data/lang_char/train_100_unigram5000_units.txt \
    --debugdir exp/train_100_pytorch_train_conformer-rnn_transducer_specaug \
    --verbose 0 \
    --resume \
    --train-json /data/dataset/asr/librispeech/dump/train_sp/deltafalse/data_unigram5000.json \
    --valid-json /data/dataset/asr/librispeech/dump/dev/deltafalse/data_unigram5000.json \
    --n-iter-processes 8
```

Note that `--ngpu` is changed to 1 for single GPU case, and `--epochs` should be 1 for single GPU performance measurement.

**Error logs**

N/A

I would highly appreciate all your feedback and comments.

**Comment by sw005320** [comment]
Thanks a lot for your investigation!
We do not actively maintain espnet1, but this seems to be an important bug fix and it would be great if you make a PR.

The espnet2 multi-gpu issues seem to depend on the environments.
Our internal investigation by @pyf98 does not show the multi-GPU issue.

**Comment by lazykyama** [comment]
@sw005320 
Thank you for the comment! Ok, let me make a PR to share current changes with the team after doing a small refactoring, and discuss more details. BTW, the overview of my modifications:
* adding a new script to launch multiple processes on each GPU like pytorch's `torch.distributed.launch`,
* adding a new option, `--use-ddp`, to `asr_train.py` to switch DDP and standard `data_parallel` when specifying multiple GPUs,
* changing evaluation schemes to distributed version,
* replacing a way to split dataset with pytorch's `DistributedSampler`,
* suppressing saving a checkpoint, intermediate info, plots, etc from non-main processes.

Regarding espnet2, I believe, when using DDP, this kind of performance slowdown won't happen. In the case of thread-based multi-GPUs, I'm not sure if the issue will happen, but it make sense that the issue depends on the environments.
I will also try to check if the performance issue happens or not in our cases.

**Comment by sw005320** [comment]
Thanks a lot!
The robust scalability is an important function (and it is difficult to manage).
So, I really appreciate your investigation.

**Comment by lazykyama** [comment]
@sw005320 sorry for my too late action... I just made a PR for this issue. I would like you to review it and give me some feedback!

BTW, regarding multi-GPU performance in v2, it looks like there is another issue around scalability. It could be possible that I made any mistakes. So, at first, let me share my investigation results below, and I would like to know your thoughts. For example, the situation is expected or not, etc. If it's actually a problem, I can open another github issue about this. Please let me know.

----

A target recipe is https://github.com/espnet/espnet/tree/master/egs2/librispeech/asr1/conf/train_asr_conformer.yaml, and the entire training times were displayed at the bottom of training log like below.

### 8 GPUs

```
[c6e67e6e69c6:0/8] 2022-05-27 07:59:22,541 (trainer:328) INFO: 2epoch results: [train] iter_time=1.824e-04, forward_time=0.065, loss_ctc=285.285, loss_att=243.501, acc=0.060, loss=256.036, backward_time=0.110, optim_step_time=0.052, optim0_lr0=2.553e-05, train_time=1.116, time=5 minutes and 8.89 seconds, total_count=2208, gpu_max_cached_mem_GB=12.502, [valid] loss_ctc=157.209, cer_ctc=0.992, loss_att=133.290, acc=0.066, cer=0.716, wer=1.000, loss=140.465, time=15.43 seconds, total_count=84, gpu_max_cached_mem_GB=12.502, [att_plot] time=51.69 seconds, total_count=0, gpu_max_cached_mem_GB=12.502
```

### 1 GPU

```
[c6e67e6e69c6] 2022-05-27 08:51:39,523 (trainer:328) INFO: 2epoch results: [train] iter_time=5.868e-04, forward_time=0.330, loss_ctc=285.325, loss_att=243.636, acc=0.060, loss=64.036, backward_time=0.303, optim_step_time=0.055, optim0_lr0=2.553e-05, train_time=3.375, time=15 minutes and 31.68 seconds, total_count=2208, gpu_max_cached_mem_GB=76.898, [valid] loss_ctc=157.512, cer_ctc=0.994, loss_att=133.760, acc=0.065, cer=0.719, wer=1.000, loss=140.886, time=19.74 seconds, total_count=86, gpu_max_cached_mem_GB=76.898, [att_plot] time=43.3 seconds, total_count=0, gpu_max_cached_mem_GB=76.898
```

In summary, each training time is:

* 8GPU: `5 minutes and 8.89 seconds`, 
* 1GPU: `15 minutes and 31.68 seconds`.

This means that 8GPU achieved only ~3x faster. It's not good result.

As far as I investigated, at least in the case of `batch_type=numel`, a scheme to split batch to each GPU should be improved. At [abs_task.py#L1524-L1533](https://github.com/espnet/espnet/blob/4806c23a002de1708c009ec6c556779115ed35b6/espnet2/tasks/abs_task.py#L1524-L1533), the batch is split into each process, but this causes smaller size of batch on each GPU.
https://github.com/espnet/espnet/blob/4806c23a002de1708c009ec6c556779115ed35b6/espnet2/tasks/abs_task.py#L1524-L1533

In fact, by adding a log around [trainer.py#L518](https://github.com/espnet/espnet/blob/4806c23a002de1708c009ec6c556779115ed35b6/espnet2/train/trainer.py#L518) like below, we can see smaller size of batch in 8 GPUs case. (e.g., "speech: 8" or "speech: 10" in 8 GPUs case vs "speech: 106" or "speech: 64" in 1 GPU case)
https://github.com/espnet/espnet/blob/4806c23a002de1708c009ec6c556779115ed35b6/espnet2/train/trainer.py#L518

```
517                 with reporter.measure_time("forward_time"):
518                     _dbg_msg = ", ".join(f"({k}: {len(v)})" for k, v in batch.items())
519                     logging.info(f"!!!! {_dbg_msg} !!!!")
520                     retval = model(**batch)
```

### 8 GPUs

```
[c6e67e6e69c6:0/8] 2022-05-27 11:11:23,884 (distributed:672) INFO: Reducer buckets have been rebuilt in this iteration.
[c6e67e6e69c6:0/8] 2022-05-27 11:11:24,363 (trainer:519) INFO: !!!! (speech: 8), (speech_lengths: 8), (text: 8), (text_lengths: 8), (utt_id: 8) !!!!
[c6e67e6e69c6:0/8] 2022-05-27 11:11:24,622 (trainer:519) INFO: !!!! (speech: 10), (speech_lengths: 10), (text: 10), (text_lengths: 10), (utt_id: 10) !!!!
[c6e67e6e69c6:0/8] 2022-05-27 11:11:24,882 (trainer:616) WARNING: The grad norm is nan. Skipping updating the model.
[c6e67e6e69c6:0/8] 2022-05-27 11:11:24,896 (trainer:519) INFO: !!!! (speech: 10), (speech_lengths: 10), (text: 10), (text_lengths: 10), (utt_id: 10) !!!!
[c6e67e6e69c6:0/8] 2022-05-27 11:11:25,144 (trainer:519) INFO: !!!! (speech: 9), (speech_lengths: 9), (text: 9), (text_lengths: 9), (utt_id: 9) !!!!
[c6e67e6e69c6:0/8] 2022-05-27 11:11:25,403 (trainer:519) INFO: !!!! (speech: 11), (speech_lengths: 11), (text: 11), (text_lengths: 11), (utt_id: 11) !!!!
[c6e67e6e69c6:0/8] 2022-05-27 11:11:25,653 (trainer:519) INFO: !!!! (speech: 8), (speech_lengths: 8), (text: 8), (text_lengths: 8), (utt_id: 8) !!!!
[c6e67e6e69c6:0/8] 2022-05-27 11:11:25,939 (trainer:616) WARNING: The grad norm is nan. Skipping updating the model.
[c6e67e6e69c6:0/8] 2022-05-27 11:11:25,963 (trainer:519) INFO: !!!! (speech: 8), (speech_lengths: 8), (text: 8), (text_lengths: 8), (utt_id: 8) !!!!
[c6e67e6e69c6:0/8] 2022-05-27 11:11:26,195 (trainer:519) INFO: !!!! (speech: 9), (speech_lengths: 9), (text: 9), (text_lengths: 9), (utt_id: 9) !!!!
[c6e67e6e69c6:0/8] 2022-05-27 11:11:26,447 (trainer:519) INFO: !!!! (speech: 9), (speech_lengths: 9), (text: 9), (text_lengths: 9), (utt_id: 9) !!!!
```

### 1 GPU

```
[c6e67e6e69c6] 2022-05-27 11:10:15,869 (trainer:519) INFO: !!!! (speech: 106), (speech_lengths: 106), (text: 106), (text_lengths: 106), (utt_id: 106) !!!!
[c6e67e6e69c6] 2022-05-27 11:10:16,523 (trainer:519) INFO: !!!! (speech: 64), (speech_lengths: 64), (text: 64), (text_lengths: 64), (utt_id: 64) !!!!
[c6e67e6e69c6] 2022-05-27 11:10:17,537 (trainer:519) INFO: !!!! (speech: 80), (speech_lengths: 80), (text: 80), (text_lengths: 80), (utt_id: 80) !!!!
[c6e67e6e69c6] 2022-05-27 11:10:18,427 (trainer:616) WARNING: The grad norm is inf. Skipping updating the model.
[c6e67e6e69c6] 2022-05-27 11:10:18,444 (trainer:519) INFO: !!!! (speech: 79), (speech_lengths: 79), (text: 79), (text_lengths: 79), (utt_id: 79) !!!!
[c6e67e6e69c6] 2022-05-27 11:10:19,257 (trainer:519) INFO: !!!! (speech: 68), (speech_lengths: 68), (text: 68), (text_lengths: 68), (utt_id: 68) !!!!
[c6e67e6e69c6] 2022-05-27 11:10:20,138 (trainer:519) INFO: !!!! (speech: 81), (speech_lengths: 81), (text: 81), (text_lengths: 81), (utt_id: 81) !!!!
[c6e67e6e69c6] 2022-05-27 11:10:20,997 (trainer:519) INFO: !!!! (speech: 64), (speech_lengths: 64), (text: 64), (text_lengths: 64), (utt_id: 64) !!!!
```

It looks like this can cause lower GPU usage in multi-GPU training, and, potentially, lower model accuracy when users want to migrate from v1 to v2 with the same hyper parameters like learning rate because learning rate is an important hyper parameter in the case of multi-GPU setting and it should be adjusted with the change of total batch size. If the batch size is implicitly changed, it's not easy for users to recognize the model accuracy issue under the same learning rate.

Note that I used following commands on DGX A100 for this investigation.

```
bash asr.sh \
    --lang en \
    --nbpe 5000 \
    --max_wav_duration 30 \
    --speed_perturb_factors "0.9 1.0 1.1" \
    --train_set "train_100" \
    --valid_set "dev" \
    --test_sets "test_clean test_other dev_clean dev_other" \
    --lm_train_text "data/train_100/text data/local/other_text/text" \
    --bpe_train_text "data/train_100/text" \
    --stage 11 \
    --stop_stage 11 \
    --skip_eval true \
    --ngpu 8 \
    --asr_config conf/train_asr_conformer.yaml \
    --lm_config conf/tuning/lm.yaml \
    --local_data_opts "--train_set train_100" \
    --asr_args "--max_epoch 2 --batch_size 32"
```

**Comment by pyf98** [comment]
Hi @lazykyama Thanks for the investigation.

In ESPnet2, did you increase the `batch_bins` with more GPUs? The config is different from that in ESPnet1, as described below:
https://espnet.github.io/espnet/espnet2_training_option.html#the-relation-between-mini-batch-size-and-number-of-gpus

In ESPnet2, with more GPUs, the batch size needs to be scaled accordingly, which seems a more typical behavior in other toolkits.

**Comment by lazykyama** [comment]
Hi @pyf98 Thank you for your comment. As you pointed me out, by changing `batch_bins` from 35000000 (default) to 280000000 (x8 bigger), the 8GPUs performance has been improved and it achieved x6.48 faster in the context of 1GPU vs 8GPUs. Sorry to bother you.

BTW, hopefully, it would be great if ESPnet2 could have an option to enable auto-tuning for such parameters because it can be automated to adjust the size of total batch, etc based on a given size per GPU. From user point of view, fewer parameters should be manually adjusted are better.

**Comment by sw005320** [comment]
> Hi @pyf98 Thank you for your comment. As you pointed me out, by changing `batch_bins` from 35000000 (default) to 280000000 (x8 bigger), the 8GPUs performance has been improved and it achieved x6.48 faster in the context of 1GPU vs 8GPUs. Sorry to bother you.

Thanks for your report!
I think this is a typical confusion.
It would be better to emphasize it in the doc.

@pyf98, could I ask you to add a remark about the different usage of the batch size to the following documents? 
https://github.com/espnet/espnet/blob/master/doc/espnet2_tutorial.md
https://github.com/espnet/espnet/blob/master/doc/espnet2_distributed.md
https://github.com/espnet/espnet/blob/master/doc/tutorial.md#multiple-gpu-tips

> BTW, hopefully, it would be great if ESPnet2 could have an option to enable auto-tuning for such parameters because it can be automated to adjust the size of total batch, etc based on a given size per GPU. From user point of view, fewer parameters should be manually adjusted are better.

I added a feature request tag to this issue for our future development items.
FYI, based on the previous experience in espnet1, this automatic adjustment was buggy.
But it's worth considering it.

**Comment by pyf98** [comment]
Thanks @lazykyama for your new investigation. The batch size issue seems quite common.

@sw005320 I have updated the three docs and made a PR here: https://github.com/espnet/espnet/pull/4436
